### DISCUSSION TITLE
How can I directly modify and view the DRAM contents?

### DISCUSSION THREAD

**Discussion by tuckertravins** [accel-sim/accel-sim-framework#172] [Q&A] [answered]
I am looking for a way to modify specific physical DRAM locations (bank, row, etc) and view the DRAM contents.
I've been printing the DRAM address accesses that get popped off the FIFO queue in dram.cc, though these only print in trace-driven mode, not PTX. Is there a good way to view DRAM contents and make modifications? Would love any advice on this or messing with DRAM in general.

**Comment by JRPan** [comment, marked as answer]
Could you please explain more? Are you trying to get the actual data in DRAM? Or do you want to get the address of each DRAM access? Or do you want to change the data access pattern?

**Reply by tuckertravins** [reply]
Sorry for not being specific. Just wanted to look at the actual data in DRAM (like what bits are high and low in a certain row) and then possibly if there is a way to modify specific locations (banks, rows). Outside of what a normal Cuda program does.

**Reply by JRPan** [reply]
I'm assuming you are using PTX mode. There is no functional model in trace mode so you cannot "read" anything. 
You can reference how load instruction is implemented to understand how to read data from memory.
https://github.com/accel-sim/gpgpu-sim_distribution/blob/dev/src/cuda-sim/instructions.cc#L3398-L3399

Thanks,
Junrui

**Reply by tuckertravins** [reply]
That makes sense, I will look into this. Thank you for great quick responses!
